egg/platform_x86.py

```diff
@@ -685,6 +685,151 @@ def store(simd_ext, typ, aligned):
+                  int i;
+                  {pre}storeu_ps(buf0, {in1}.v0);
+                  {pre}storeu_ps(buf1, {in1}.v1);
+                  for (i=0; i<{nelts2}; ++i) {{
+                    if (mask0 & (1U << i)) {{
+                      {in0}[i] = nsimd_f32_to_f16(buf0[i]);
+                    }}
+                  }}
+                  for (i=0; i<{nelts2}; ++i) {{
+                    if (mask1 & (1U << i)) {{
+                      {in0}[i+{nelts2}] = nsimd_f32_to_f16(buf1[i]);
+                    }}
+                  }}'''.format(nelts2=nelts2, **fmtspec)
+    if simd_ext in sse + avx and typ[0] == 'f':
+        r = ['''unsigned mask = {pre}movemask{suf}({in2});
+                {typ} buf[{nelts}];
+                {pre}storeu{suf}(buf, {in1});'''.format(nelts=nelts, **fmtspec)]
+        r += ['''if (mask & {mask}U) {{
+                   {in0}[{i}] = buf[{i}];
+                 }}'''.format(i=i, mask=1<<i, **fmtspec)
+              for i in range(0, nelts)]
+        return '\n'.join(r)
+    if simd_ext == 'avx2' and typ[0] != 'f':
+        bitratio = tbits // 8
+        return '''unsigned mask = {pre}movemask_epi8({in2});
+                  {typ} buf[{nelts}];
+                  int i;
+                  {pre}storeu_si{nbits}((__m{nbits}i*)buf, {in1});
+                  for (i=0; i<{nelts}; ++i) {{
+                    if (mask & (1U << ({bitratio}*i+{bitratio}-1))) {{
+                      {in0}[i] = buf[i];
+                    }}
+                  }}'''.format(nelts=nelts, bitratio=bitratio, **fmtspec)
+    if simd_ext in avx512:
+        if typ[0] == 'f':
+            store_buf = '{pre}storeu{suf}(buf, {in1});'.format(**fmtspec)
+        else:
+            store_buf = '{pre}storeu_si{nbits}(buf, {in1});'.format(**fmtspec)
+        return '''{typ} buf[{nelts}];
+                  {store_buf}
+                  int i;
+                  for (i=0; i<{nelts}; ++i) {{
+                    if ({in2} & (1ULL << i)) {{
+                      {in0}[i] = buf[i];
+                    }}
+                  }}'''.format(store_buf=store_buf, nelts=nelts, **fmtspec)
+    # integer mask is not available
+    if simd_ext in sse + avx:
+        if typ[0] == 'f':
+            store_buf = '''{pre}storeu{suf}(buf, {in1});
+                           {pre}storeu{suf}(mask, {in2});'''.\
+                        format(**fmtspec)
+        else:
+            store_buf = \
+                '''{pre}storeu_si{nbits}((__m{nbits}i*)buf, {in1});
+                   {pre}storeu_si{nbits}((__m{nbits}i*)mask, {in2});'''.\
+                format(**fmtspec)
+        return '''{typ} buf[{nelts}], mask[{nelts}];
+                  int i;
+                  {store_buf}
+                  for (i=0; i<{nelts}; ++i) {{
+                    if (mask[i]) {{
+                      {in0}[i] = buf[i];
+                    }}
+                  }}'''.format(store_buf=store_buf, nelts=nelts, **fmtspec)
+    assert False
+
```

Comment:
I originally structured the code by types and sizes, and found that this leads to much code duplication, because the way the various SSE/AVX instructions handle types and sizes is not logical at all. That is, I didn't find a logical division such as "for 32- and 64-bit types, do this; for 8- and 16-bytes, do that", which didn't duplicate code.

The code is now structured by instruction set, from most specific to most generic, and each `if` statement picks what it can handle and ignores the rest, which falls through to the next `if` statement. The final `assert` handles logic errors that shouldn't occur. The order of the `if` blocks could be changed at will, and I ordered them to ensure the most efficient cases are checked first.

I will refactor the code for clarity. I will see whether it make sense to add explicit statements for 8- or 16-bit elements.